README


This archive contains the distribution Backup-Email, version
0.022:

  A backup application specifically built for backing up files to an IMAP server.

This software is copyright (c) 2009 by Stefan Petrea.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension Backup::Email.

0.02   Fri Nov 13 19:00:00 2009
        - Added some more tests to Backup::Email::Config and g(). 
        - Moved to Dist::Zilla for building the distribution ( not sure I'm supposed to write this as a change...)

0.015  Wed Oct  7 18:29:00 2009
        - added support in Backup::Email::Send role for attaching multiple files to an email
        - changed all hardcoded paths to appdir path taken from config.yml as documented in Backup::Email
        - fixed some typos
               
0.01  Fri Oct  2 15:50:29 2009
        - original version; created by h2xs 1.23 with options
                -XA -n Backup::Email



Makefile.PL

use strict;
use warnings;

use ExtUtils::MakeMaker;

WriteMakefile(
  DISTNAME  => 'Backup-Email',
  NAME      => 'Backup::Email',
  AUTHOR    => "Stefan\ Petrea\ \<stefan\.petrea\@gmail\.com\>",
  ABSTRACT  => "A\ backup\ application\ specifically\ built\ for\ backing\ up\ files\ to\ an\ IMAP\ server\.",
  VERSION   => '0.022',
  EXE_FILES => [ qw() ],
  (eval { ExtUtils::MakeMaker->VERSION(6.31) } ? (LICENSE => 'perl') : ()),
  PREREQ_PM    => {
    "Alien::wxWidgets" => '0',
    "Wx" => '0',
    "DateTime" => '0',
    "Mail::IMAPClient" => '0',
    "Net::SSLeay" => '0',
    "DBM::Deep" => '0',
    "YAML" => '0',
    "Archive::Zip" => '0',
    "Net::SMTP::TLS" => '0',
    "MooseX::NonMoose" => '0',
    "List::Util" => '0',
    "MIME::Lite" => '0',
    "List::MoreUtils" => '0',
    "Moose" => '0',
    "IO::Socket::SSL" => '0',
    "List::AllUtils" => '0',
    "perl" => '5.010000',
    "MIME::Base64" => '0',
  },
  test => {TESTS => 't/*.t'}
);



lib/Backup/Zip.pm
# 
# This file is part of Backup-Email
# 
# This software is copyright (c) 2009 by Stefan Petrea.
# 
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
# 
use strict;
use warnings;
package Backup::Zip;
use Moose::Role;
use Archive::Zip qw/:ERROR_CODES :CONSTANTS/;
use Carp qw/confess/;
use 5.008;
use strict;
use warnings;


requires 'files_to_zip';


=pod


=head1 NAME

Backup::Zip - A Zip Role for classes using Moose


=head1 VERSION

version 0.022

=head1 SYNOPSIS

        package CanZipThings;
        use Moose;
        with 'Backup::Zip';

        sub files_to_zip() {
            return (
                '/etc/passwd',
                '/etc/bashrc',
            );
        }

        package main;
        use CanZipThings;

        my $canzip = CanZipThings->new;
        $canzip->make_zip("/tmp/somezip.zip");

=head1 DESCRIPTION

This is a Role and it provides a zip function to the consumer class.

=head1 files_to_zip()

Backup::Zip is a Role and it depends on the files_to_zip() method which it expects the
consumer class to provide.
This method will be used by make_zip() in order to get an array of the files it will zip
so the return value should be a list with filepaths.

=head1 make_zip($zip)

This method uses Archive::Zip to zip your files and stores them in the path indicated by $zip.
If you provide no arguments , it will 'confess' with an error message.
You will also receive warnings for all the files that don't exist.

=head1 unzip($what,$where)  - unimplemented!

This method will unzip the files found at the path $what to the path $where.

=head1 BUGS

If this module has any bugs please report them to rt.cpan.org or my email.

Also if you have any suggestions on the code you are welcome to drop me an e-mail.

=head1 AUTHOR

Stefan Petrea, C<< <stefan.petrea at gmail.com> >>

=cut





sub make_zip {
	my ($self,$zip) = @_;
	confess 'missing zip filename as argument' unless $zip;
	my $zip_obj = Archive::Zip->new();
	for my $file ( $self->files_to_zip ) { # $file can be a directory also
		if	( -e $file ) {
			$zip_obj->addTree($file,'');
		} else {
			warn "File not found -> ".$file; 
		};
	};
	if ( $zip_obj->writeToFileNamed($zip) != AZ_OK ) {
		confess 'cannot write zip file';
	}
}

1;

lib/Backup/Email/GUI.pm
# 
# This file is part of Backup-Email
# 
# This software is copyright (c) 2009 by Stefan Petrea.
# 
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
# 
use strict;
use warnings;
package Backup::Email::GUI;
use strict;
use warnings;
use Moose;
use MooseX::NonMoose; # for inheriting from Non-Moose classes
use Backup::Email;
#use feature 'say';
use Data::Dumper;
use DateTime;
use Archive::Zip;
use List::MoreUtils qw(each_arrayref);
# Wx::Perl::* is the namespace for a lot of Wx controls

use Wx qw(:listctrl wxRED wxBLUE wxITALIC_FONT wxDefaultPosition wxDefaultSize);
use Wx::Event qw( EVT_BUTTON);
eval { require 'Wx::ListCtrl' };

extends 'Wx::App';

=pod


=head1 NAME

Backup::Email::GUI - This is a GUI written using the Wx bindings for Backup::Email.

=head1 VERSION

version 0.022

=head1 SCREENSHOT

=begin html

<img src="http://perlhobby.googlecode.com/svn/trunk/scripturi_perl_teste/backup_pl/GUI.png" />

=end html

=head1 map_multiple($code,@cols)

Extends the Perl 'map' operator to iterating over multiple lists at the same time.

Parameters

=over

=item * $code

This is the code that is going to be run, $code is a subref. $code will receive a list of arguments formed by getting one argument
from each of the arrayrefs in @cols.

=item * @cols

@cols is an list of arrayrefs.

=back


=head1 BUGS

If this module has any bugs please report them to rt.cpan.org or my email.

Also if you have any suggestions on the code you are welcome to drop me an e-mail.

=head1 AUTHOR

Stefan Petrea, C<< <stefan.petrea at gmail.com> >>

=cut




# TODO -> make these constants
# a couple of constant IDs needed for Wx widgets
my $RESTORE_ID = 3;
my $BACKUP_ID    = 4;
my $LISTCTRL_ID = 8;

has backup => (
	isa	=> 'Any',
	is	=> 'rw',
	lazy	=> 1,
	default	=> sub {
		my $b = Backup::Email->new('/home/spx2/config.yml');
		$b->check_if_connected;
		$b;
	},
);

has frame => (
	isa	=> 'Any',
	is	=> 'rw',
	lazy	=> 1,
	default	=> sub {
		print "frame created\n";
		return Wx::Frame->new( undef,         # Parent window
			-1,            # Window id
			'Email Backup', # Title
			[400,100],         # position X, Y
			[850,400]     # size X, Y
		);
	}
);

has list => (
	isa	=> 'Any',
	is	=> 'rw',
	lazy	=> 1,
	default	=> sub {
		my $self = shift;
		return Wx::ListCtrl->new( 
			$self->frame, 
			$LISTCTRL_ID, 
			wxDefaultPosition, 
			[800,300] , 
			wxLC_REPORT 
		);
	}
);


sub RESTORE {
	my $self = shift;
	print "RESTORE pressed\n";
	#say Dumper $self->GetSelectedItem;
	my $selected = $self->GetSelectedItem;
	if( defined $selected ) {
            # Cwd::abs_path, File::Spec->rel2abs , portable way  to turn relative -> absolute path 
            my $where_to_extract = $self->backup->appdir.'/backup_pl/unpacked/';
            `mkdir $where_to_extract`;
            `rm $where_to_extract/*`;
            my $uid = $self->list->GetItem($selected,0)->GetText;
            my $raw_data = $self->backup->getMessageFile($uid);
            # this path should also be relative to the application path $self->appdir (we leave it like that testing purposes)
            my $archive = Archive::Zip->new('config.zip');
            $archive->extractTree('',$where_to_extract);
	} else {
            warn "[WARNING] No message was selected to restore from"; #
        };
}

sub BACKUP {
	my $self = shift;
        confess "[ERROR] You need a ID in your configuration YAML file" unless $self->backup->ID;
        my $zipfile = 'configs.zip';
        `rm $zipfile`;
        print Dumper $self->backup->files_to_zip()."\n";
        $self->backup->make_zip($zipfile);

        $self->backup->attachFile($zipfile);
        $self->backup->sendEmail({
                subject => 
                    'backup '.$self->backup->ID.' '.DateTime->now.' OS='.$^O.' [configuration]' ,
                body    => join("\n",$self->backup->files_to_zip()),
        });
	print "BACKUP pressed\n";
}


sub GetSelectedItem {
	my ( $self ) = @_;  
	# find selected items
	my $item = -1;
	while(1) {
		$item = $self->list->GetNextItem(
			$item,      
			wxLIST_NEXT_ALL,
			wxLIST_STATE_SELECTED
		);
		last if(-1 == $item);
		return $item;
	};
	return undef;
}


sub map_multiple {
	my ($self,$code,@cols) = @_;
	my $columns = each_arrayref(@cols);
	while( my @pair = $columns->() ) {
		$code->(@pair);
	}
}


sub OnInit {
	my $self = shift;

	my $row_id = 1;
	#my $wlc = $self->frame->{LISTCTRL}; = ;



	$self->frame->{ RESTORE } = Wx::Button->new($self->frame, $RESTORE_ID, 'RESTORE' , [40,300] );
	EVT_BUTTON( $self->frame , $RESTORE_ID , sub { $self->RESTORE } );


	$self->frame->{ BACKUP } = Wx::Button->new($self->frame, $BACKUP_ID, 'BACKUP' , [130,300] );
	EVT_BUTTON( $self->frame , $BACKUP_ID , sub { $self->BACKUP } );



	#my $columns = each_arrayref(
		#[0..3],
		#[qw/uid	From	Subject	Date/],
		#[qw/80	110	400	180/],
	#);
	#while( my ($idx,$name,$width) = $columns->() ) {
		#$self->list->InsertColumn($idx, $name );
		#$self->list->SetColumnWidth($idx,$width);
	#}

	$self->map_multiple( 
		sub {
			$self->list->InsertColumn($_[0], $_[1] );
			$self->list->SetColumnWidth($_[0],$_[2]);
		},
		[0..3], # $_[0]
		[qw/uid	From	Subject	Date/], # $_[1]
		[qw/80	110	400	180/], # $_[2]
	);




	$self->backup->check_folder('INBOX',
		sub {
			my ($s,$hashref,$uid) = @_; # From Subject Date


                        $hashref = ${ $hashref };
                        #say Dumper $hashref;
                        # took out the shift because it would modify the DBM::Deep on disk as well
			my $from	= @{$hashref->{'From'}}[0];
                        my $subject	= @{$hashref->{'Subject'}}[0];
                        my $date	= @{$hashref->{'Date'}}[0];

			my $entry_id = $self->list->InsertStringItem( $row_id , 2 );
			$self->list->SetItemData( $entry_id, int rand 100 );

			#$self->list->SetItem($entry_id,0,"$uid");
			#$self->list->SetItem($entry_id,1,$from);
			#$self->list->SetItem($entry_id,2,$subject);
			#$self->list->SetItem($entry_id,3,$date);

			$self->map_multiple(
				sub {
                                    #local($_[0]);
                                    $self->list->SetItem($entry_id,$_[0],$_[1]);
				},
				[0..3],
				["$uid",$from,$subject,$date]
			);


			#print "\n$uid\n";
			$row_id++;
		}
	);

	$self->SetTopWindow($self->frame);    # Define the toplevel window
	$self->frame->Show(1);                # Show the frame
	1;
}




1;

lib/Backup/Email/Send.pm
# 
# This file is part of Backup-Email
# 
# This software is copyright (c) 2009 by Stefan Petrea.
# 
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
# 
use strict;
use warnings;
package Backup::Email::Send;
use Moose::Role;
use Carp qw/confess/;
use MIME::Lite;
use List::AllUtils qw/all/;
use Net::SMTP::TLS;
use Data::Dumper;
=pod

=head1 NAME

Backup::Email::Send - A role for sending e-mails.

=head1 VERSION

version 0.022

=head1 DESCRIPTION

This role will provide sending e-mails capabilities to the class that consumes it.
It's requirements are either the class that consumes the role is Backup::Email::Config or it provides methods:

=over

=item * to()

=item * from()

=item * smtp()

=item * username()

=item * password()

=back

=head1 attachFile($path)

Adds a given existing file to the list of files which will be attached to the e-mail

=head1 sendEmail($href)

$href is HashRef which contains 2 keys: subject,body.
It must be filled if you want to have those two attributes for the e-mail you are sending.
Also , here , the list of files are checked and attached to the e-mail message.

=head1 TODO

It should also allow for attributes instead of the methods.

=head1 BUGS

If this module has any bugs please report them to rt.cpan.org or my email.

Also if you have any suggestions on the code you are welcome to drop me an e-mail.

=head1 AUTHOR

Stefan Petrea, C<< <stefan.petrea at gmail.com> >>

=cut

#
# we are using functions generated by AUTOLOAD, it would be nice
# if we could do a require on ->to,->from,->smtp,->username,->password it would be
# very good , so that we can make at least a check for being able to compose this
# role with a particular class
#


has attachments => (
    isa => 'ArrayRef[Str]',
    is  => 'rw',
    default => sub {[]},
);


sub BUILD {
    my $self = shift;
    confess qq/
    [ERROR]
    The class that consumes Backup::Email::Send needs to inherit
    from Backup::Email::Config as well or at least provide all of the following methods
    ->to , ->from , ->username , ->password , ->smtp
    / unless (
        $self->isa('Backup::Email::Config')
        ||
        all { $self->does($_) }
            qw/to from username password smtp/
    );
};




sub attachFile {
	my ($self,$path) = @_;
        confess "file $path to be attached does not exist" unless -f $path;
        push @{$self->attachments},$path;
}

sub sendEmail {
	my ($self,$href) = @_;

	my $msg = MIME::Lite->new(
		From	=> "".$self->to,
		To	=> "".$self->from,
		Subject	=> $href->{subject},
                Type    => 'TEXT',
                Data    => $href->{body},
	);


        warn "no file attachments" unless @{$self->attachments};

        -f $_
        ?   $msg->attach(
		Type	=> 'application/octet-stream',
		Path	=> $_,
            )
        : warn "file $_ does not exist"
            for @{$self->attachments};

	my $smtp_tls =  MIME::Lite::SMTP::TLS->new(
		"".$self->smtp,
		User		=> "".$self->username,
		Password	=> "".$self->password,
		Port		=> 587,
	);

	$smtp_tls->mail("".$self->from);
	$smtp_tls->to("".$self->to);

        print Dumper $msg;
	$smtp_tls->data();
	$msg->print_for_smtp($smtp_tls);
	$smtp_tls->dataend();
        # here need to check if e-mail has been sent, otherwise throw exception

        $self->attachments([]);#erase all attachments if email has been sent succesfuly
}


# monkeypatching a.k.a make new 'fake' package to inject a new method in it ( ::print is needed by $msg->print_for_smtp )
@MIME::Lite::SMTP::TLS::ISA = qw( Net::SMTP::TLS );
sub MIME::Lite::SMTP::TLS::print { shift->datasend(@_) }

1;

lib/Backup/Email/Config.pm
# 
# This file is part of Backup-Email
# 
# This software is copyright (c) 2009 by Stefan Petrea.
# 
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
# 
use strict;
use warnings;
package Backup::Email::Config;
#this package will be used to refactor the use of the configuration file
#use Want;
use Moose;
use strict;
use warnings;
use YAML qw/LoadFile DumpFile/;
use Data::Dumper;
use Carp qw/confess/;
#use MooseX::Types::Moose qw/HashRef Str/;
#use feature 'say';
use overload 
	'@{}'	=> \&arrayify,
	'""'	=> \&stringify;
our $AUTOLOAD;

=pod

=head1 NAME

Backup::Email::Config - A module that maps naturally keys a YAML file over method names

=head1 VERSION

version 0.022

=head1 DESCRIPTION

This module was written to map every item from an YAML to a chain of calls of methods
of an object.
It's intended to ease the use of a YAML.


I<quote>:
C<< <@nothingmuch> it's kinda like making an omlette with a remote control >>


=head1 CONFIGURATION FILE

	ID: <id_of_machine>
	from: <some_address>
	imap: <imap_server>
	password: <yourpassword>
	smtp: <your_smtp_server>
	to: <address_you_want_to_send_to>
	username: <username>
	file_list:
		- c:\Perl\stuff
		- c:\Program Files\Vim\vim72\_vimrc


The configuration file -> config.yml should be located in the same directory with app.pl.

The desired behaviour is the following:

=head1 SYNOPSIS

Having an object $c of type Backup::Email::Config ( the naming should probably change to something more general )
and an YAML file for example:

        ---
        ID: man
        animal: pig
        object:
              - pq
              - qp
              - test
        one:
              two:
                      three: thrice

We list the following calls:

=begin html

<table border=1>
<tr>
<th> CALL				</th><th> RETURN VALUE 		</th>
</tr>

<tr>
</td><td>$c->ID 				</td><td> 'man'			</td>
</tr>

<tr>
</td><td>$c->animal				</td><td> 'pig'		 	</td>
</tr>

<tr>
</td><td>$c->object				</td><td> ('pq','qp','test')	</td>
</tr>

<tr>
</td><td>$c->object(0)			</td><td> 'pq'			</td>
</tr>

<tr>
</td><td>$c->object(1)			</td><td> 'qp'			</td>
</tr>

<tr>
</td><td>$c->one				</td><td> {two=>{three=>thrice}}</td>
</tr>

<tr>
</td><td>$c->one->two			</td><td> {three=>thrice}	</td>
</tr>
</table>

=end html

=head1 g()

Takes as arguments a set of strings/numbers to indicate keys/indexes to the configuration data.
Example
            $config->g qw/object 1/

This should return the string 'qp' for the above config file.

=head1 HOW IT WORKS

We're using AUTOLOAD to catch undefined method calls , and we strip them off of their namespace, and we push them in an array(chain).

AUTOLOAD always returns $self so this allows for chaining of the undefined methods.

We overload the arrayify and stringify subs. (arrayify is just a proxy to stringify)

In stringify we need to get our array(chain) and step-by-step get to the data we need.

So if in the chain we have ('one','two','three') our expression will look actually access $self->_config->{one}->{two}->{three},

but we just called $c->one->two->three which is pretty natural.

=head1 TODO

If we have the chain $o->m1->m2->m3->m4;
Is there any way Want.pm can be used to distinguish the call that is inside a chain(i.e. m1,m2,m3) and the one at the end of the
chain(m4) ?
This would allow to avoid tricks like ''.$config->m1->m2



=cut

has _config => (
	isa	=> 'HashRef',
	is	=> 'rw',
);

has _config_path => (
	isa	=> 'Str',
	is	=> 'rw',
);


has chain	=> (
	isa	=> 'ArrayRef',
	is	=> 'rw',
	default	=> sub{[]},
);

has last_config_item => (
	isa	=> 'Str',
	is	=> 'rw',
	default	=> '',
); # last config item resulted from a chain


has temp_change => ( # if change is part of the chain last_config_item needs to be passed to him also
	isa	=> 'Str',
	is	=> 'rw',
	default	=> '',
); # last config item resulted from a chain



sub BUILDARGS {
	my ($self,$file) = @_;
	confess
	'config parameter invalid and default config.yml path invalid' if !defined($file) && !(-e $self->appdir.'/config.yml');
	{
		_config_path	=> $file// $self->appdir.'/config.yml' 
	};
};

#after 'new' => {
sub BUILD {
	my ($self) = @_;
	$self->_config( LoadFile($self->_config_path) );
}

sub reload {
	my ($self) = @_;
	$self->_config( LoadFile($self->_config_path) );
}


#
# Here we just take any method applied to this class and associate it to some
# value in the YAML ( accessors on the fly )
#



sub AUTOLOAD {
	# this is a link in the chain
	
	# will also be called if you misspell an existing method ... no cure for that , unless it's not in the YAML either and then
	# some kind of warning/exception should pop up , will have to implement this
	my ($self,$index) = @_;
	#say "AUTOLOAD: $AUTOLOAD";
	# should verify that AUTOLOAD starts with Backup::Email::Config , otherwise return to 'normal' flow
	$AUTOLOAD =~ s/^.*::(\w*)$/$1/; # actually , we don't need the rest
	push @{$self->chain()} , [ $1 , $index ];
        if(wantarray()){
            #say 'LIST';
             my $ref = $self->arrayify;
             my @res = eval { @{$ref} };
             if($@) { @res = %{$ref} };
             return @res;
        };
#        if(want(qw/LVALUE SCALAR/)){
#            #say 'SCALAR';
#            rreturn "$self";
#        };

        return $self;
};

sub arrayify {
    my ($self) = @_;
    #say "DEBUG: ARRAYIFY is CALLED!";
    
    # this is for this particular case   my @stuff = <Backup::Email::Config>->m1->m2->m3;
    # it will call arraify because Backup::Email::Config is a blessed hash that needs to be
    # dereferenced in order to be converted to an array , however we overload this
    # let stringify take care of cleaning stuff up also
    return $self->stringify;
}


sub stringify {
	my ($self) = @_;
	#need to update this logic to the context of stringify
	#say "DEBUG: STRINGIFY is CALLED!";
	confess "chain empty inside stringify()" unless @{ $self->chain };
	my $return = $self->_config; # $return will be a HASH throughout the following loop and at the end a SCALAR or ARRAYREF
	for my $call ( @{$self->chain} ) {
		my ($name,$index) = @$call;
		if	( defined $index ) {
			$return = $return->{$name}->[$index];
			$self->last_config_item( $self->last_config_item . "->{$name}->[$index]" );
		} elsif	( defined $name ) {
			$return = $return->{$name} if ref($return) eq 'HASH';
			$self->last_config_item( $self->last_config_item . "->{$name}" );
		}
	};
	$self->temp_change($self->last_config_item);
	$self->last_config_item('');
	$self->chain([]); # clean the chain
	return $return;
}


sub debug {
	my ($self) = @_;
	print "last_config_item	: ".$self->last_config_item."\n";
	print "temp_change	: ".$self->temp_change."\n";
}

sub change { # append this to the end of the chain to modify and save the conf file
	my ($self,$what) = @_;

	"$self"; #force stringify to be called to get temp_change right because if we append change to the chain stringify won't be called any more	 
	# will cause a warning
	my $expr = '$self->_config'.$self->temp_change.'=$what;';


	eval $expr; # we modify our $self->_config
	DumpFile($self->_config_path,$self->_config); # save the changes on disk
	$self->last_config_item(''); # clear the last_config_item
	$self->temp_change('');
}


sub g {
    my $self = shift;
    my $return = $self->_config;
    for my $arg (@_) {
        for(ref($return)){
            /HASH/  and $return = $return->{$arg} and next;
            /ARRAY/ and $arg=~/^\d+$/ and $return = $return->[$arg]
                or confess __PACKAGE__.": argument error in g() , expected numeric and got [$arg]";
            #confess "wrong arguments to g() [$@]" if $@;
        }
    };
    return $return;
}



=head1 SEE ALSO

If you're interested in similar modules, take a look at 

L<Data::AsObject> , L<Data::Hive> , L<MooseX::YAML>

=head1 BUGS

If this module has any bugs please report them to rt.cpan.org or my email.

Also if you have any suggestions on the code you are welcome to drop me an e-mail.

=head1 AUTHOR

Stefan Petrea, C<< <stefan.petrea at gmail.com> >>


=cut

1;


lib/Backup/Email.pm
# 
# This file is part of Backup-Email
# 
# This software is copyright (c) 2009 by Stefan Petrea.
# 
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
# 
use strict;
use warnings;
# Every program attempts to expand until it can read mail. Those programs which cannot so expand are replaced by ones which can. -> jwz
package Backup::Email;
use strict;
use warnings;
use Moose;
use Mail::IMAPClient;
use Mail::IMAPClient::BodyStructure;
use IO::Socket::SSL;
use Data::Dumper;
use MIME::Base64 qw/decode_base64/; #doesn't work properly on Windows , produces invalid 7z and characters outside base64 set
use List::Util;
use List::MoreUtils;
use DBM::Deep;
use 5.010000;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Backup::Email ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.015';




=pod

=head1 NAME

Backup::Email - A backup application specifically built for backing up files to an IMAP server.

=head1 VERSION

version 0.022

=head1 DESCRIPTION


This module coupled with Backup::Email::GUI is intended to help you backup configuration
files on your system( as well as other files that you want to specify in the configuration YAML file ).

It receives functionality for sending e-mails from the Backup::Email::Send role.

It inherits Backup::Email::Config so it can access easily the config.yml that you need to provide.

It caches some e-mail headers in messages.db using DBM::Deep so it will run faster on sub-sequent runs.

(Normally you should use something like revision control or just ftp but this is probably )

=head1 SYNOPSIS

=over 

=item config.yml file 

    The configuration file needs to specify the credentials of the e-mail and a list of files to back up.
    Your config file should be named 'config.yml' and placed in your appdir directory.

        ---
        ID: toshiba
        file_list:
          - /home/user/.muttrc
          - /etc/passwd
          - /home/user/.opera/
        imap: imap.gmail.com
        smtp: smtp.gmail.com
        appdir: /path/to/where/config.yml/is
        from: <from-email>
        to: <to-email>
        password: <your_password>
        username: <your_username>

=back



=head1 client

Mail::IMAPClient object used to log-in

=head1 getMessageFile($uid)

This method will get the attached zip file for the message with uid = $uid , decode it base64( because it's supposedly coded base64 if it's been sent by Backup::Email ) and store it on disk.

=head1 check_if_connected()

Checks if we are connected, otherwise exits immediately.

=head1 check_folder($folder,$code)

Gets all messages sent by Backup::Email to the e-mail address and runs the the subref $code on each of them.
The $code sub receives the following parameters:

=begin html

<ul>
<li>$self of Backup::Email object
<li>hashref with keys 'Date' , 'Subject' , 'From'
<li>uid
</ul>

=end html 

It also uses DBM::Deep to store already fetched message headers(in messages.db) from your e-mail so it will load up faster on sub-sequent runs.

=head1 files_to_zip()

This method provides the list of files that need be zipped. This method is used by the make_to_zip() provided by the role Backup::Zip.

=cut


with 'Backup::Zip'; # this role makes the assumption that the Class consuming it has a ->files_to_zip method available
with 'Backup::Email::Send';
extends 'Backup::Email::Config'; # that way we'll get configuration stuff 'for free'

has socket => (
	isa	=> 'Any',
	is	=> 'rw',
	lazy	=> 1,
	default	=> sub {
		my ($self) = @_;
		IO::Socket::SSL->new(
			PeerAddr => "".$self->imap,
			PeerPort => 993,
		)
			or die "socket(): $@"
	},
);



has client => (
	isa	=> 'Any',
	is	=> 'rw',
	lazy	=> 1,
	default	=> sub {
		my ($self) = @_;
		Mail::IMAPClient->new(
			Socket	=> $self->socket,
			User	=> "".$self->username, # we need stringification
			Password=> "".$self->password,
			Uid	=> 1,
		)
			or die "new(): $@"
	},
);

sub check_if_connected {
	my ($self) = @_;

	if($self->client->IsAuthenticated() ){
		print "Authenticated \n";
	} else {
		print "Authentication failed \n";
		exit(-1);
	};
        #my @folders = $self->client->folders();
	#print join("\n* ", 'Folders:', @folders), "\n";
}


sub check_folder {
	my ($self,$folder,$code) = @_;
        my $dbm_file = 'messages.db';

        my $not_first_time =-f $dbm_file;
        my $db = DBM::Deep->new(
            file=>  $dbm_file,
            type => DBM::Deep->TYPE_ARRAY
        );

        # TODO -> $self->last_message_read will be replaced by just getting the first(if there is one) uid from message.db ($db)

	$self->client->select($folder);
	# pt parse_headers e by default peek pe 1 dar pt restul e pe 0
	my @uids = $self->client->search('SUBJECT "[configuration]"'); #->recent

        my $last_uid = 0;#biggest uid in the first loop
        # processing fresh uids


        print "FIRST:".$uids[0]."\n";
        if( !defined($self->last_message_read) || $uids[-1] > int($self->last_message_read) ) {
            for my $uid ( @uids ) {
                my $hashref = $self->client->parse_headers($uid,"Date","Subject","From");

                next if (
                    index( $hashref->{"Subject"}->[0] , $self->ID )<0
                    || !$uid
                    || ( defined($self->last_message_read) && $uid <= int($self->last_message_read) )
                );
                #say "UID=$uid"." ".int($self->last_message_read);
                $db->unshift({
                        uid => $uid,
                        %$hashref
                });
                $code->($self,\$hashref,$uid);
                $last_uid = 
                    $last_uid < $uid 
                    ? $uid 
                    : $last_uid;
            };
        };
        $self->last_message_read->change($last_uid) if $last_uid;

        print "!!!starting to use cache\n";
        # processing the cache
        if( $not_first_time ) {
            for my $i ( 0..$db->length() ) {
                my $hashref = $db->get($i);
                next if ($last_uid && $hashref->{uid}+1 > $last_uid );
                print "UID=".$hashref->{uid}."\n";
                $code->( $self, \$hashref, $hashref->{uid} );
            }
        }
}


sub getMessageFile {
	my ($self,$uid) = @_;
	print "in getMessage\n";

	my $struct = Mail::IMAPClient::BodyStructure->new( 
		$self->client->fetch($uid, "bodystructure")
	);

	# say Dumper $struct; 
	unlink "config.base64";
	unlink "config.zip";
	#have to remove these two on SIGQUIT or Wx close event

	
        if( $^O =~ /MSWin32/ ) {
	    open my $config_archive , ">config.base64";
            # here we get the second attachment , normally we should enumerate
            # the parts and select the one who has the content type of an archive
            # and after that docode and store on disk
            # probably will need unlink, rename, File::Copy for various operations with the files
            print $config_archive $self->client->bodypart_string($uid,2);
            #TODO -> try on windows to use  binmode FILEHANDLE nad this should fix it and eliminate the need for base64.exe
            system "base64 -d -n config.base64 > config.zip"; # this is not needed on Linux , where base64 from MIME::Base64 works properly
        } elsif ( $^O =~ /linux/ ) {
            print "I'm on linux\n";
	    open my $config_archive , ">config.zip";
            print $config_archive decode_base64( $self->client->bodypart_string($uid,2) );
        };
}



# needed for Zip Role composition
sub files_to_zip {
	my $self = shift;
        confess "[ERROR] no files to zip in configuration file" unless $self->file_list;
	my @return = @{$self->file_list};
	push @return,$self->appdir.'/config.yml';
	return @return;
}


=head1 GUI

The Backip::Email comes with a GUI which you can use by running using app.pl

=begin html

<img src="http://perlhobby.googlecode.com/svn/trunk/scripturi_perl_teste/backup_pl/GUI.png" />

=end html

=head1 NOTES

Normally you would use some version control for backing things up like this but I decided to write the module anyway.

=head1 BUGS

You can use the CPAN Request Tracker http://rt.cpan.org/ and submit
new bugs under

  http://rt.cpan.org/Ticket/Create.html?Queue=Backup::Email


=head1 SEE ALSO

L<Backup::Email::GUI> , L<Backup::Email::Config> , L<Backup::Zip> , L<Backup::Email::Send>

=cut

1;